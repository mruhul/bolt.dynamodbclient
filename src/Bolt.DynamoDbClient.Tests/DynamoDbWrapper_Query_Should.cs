using System.Net;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Bolt.Common.Extensions;
using Bolt.DynamoDbClient.Tests.Builders;
using Bolt.DynamoDbClient.Tests.TestHelpers;
using Shouldly;

namespace Bolt.DynamoDbClient.Tests;

public class DynamoDbWrapper_Query_Should
{
    private IDynamoDbWrapper sut;
    private IAmazonDynamoDB fake;

    public DynamoDbWrapper_Query_Should()
    {
        fake = Substitute.For<IAmazonDynamoDB>();
        sut = DynamoDbWrapperBuilder.Build(fake);
    }

    [Fact]
    public async void Generate_correct_input_for_equal_search()
    {
        QueryRequest? gotRequest = null;
        fake.QueryAsync(Arg.Any<QueryRequest>(), Arg.Any<CancellationToken>())
            .Returns(new QueryResponse 
            { 
                Count = 1,
                HttpStatusCode = System.Net.HttpStatusCode.OK,
                ScannedCount = 1,
                Items = new List<Dictionary<string, AttributeValue>>
                {
                    new Dictionary<string, AttributeValue>
                    {
                        ["PK"] = new AttributeValue
                        {
                            S = "book-1"
                        },
                        ["SK"] = new AttributeValue
                        {
                            S = "details"
                        },
                        ["Title"] = new AttributeValue
                        {
                            S = "book-title"
                        },
                        ["Price"] = new AttributeValue
                        {
                            N = "100.34"
                        },
                        ["Identifier"] = new AttributeValue
                        {
                            S = "8359FC3F-B5EE-4622-9993-2B5CFAE03A1E"
                        },
                        ["PublishedAt"] = new AttributeValue
                        {
                            S = "2022-12-31T13:00:00Z".ToUtcDateTime().FormatAsUtc()
                        },
                        ["PublishStatus"] = new(){
                            S = "Draft"
                        }
                    }
                }
            })
            .AndDoes(c => gotRequest = c.Arg<QueryRequest>());

        var gotRsp = await sut.Query()
            .PartitionKey().Equals("partition-key")
            .SortKey().Equals("sort-key")
            .Fetch<BookDbRecord>();

        new
        {
            gotRsp,
            gotRequest
        }.ShouldMatchApproved();
    }

    [Fact]
    public async Task work_for_reserved_words()
    {
        var fakeDb = Substitute.For<IAmazonDynamoDB>();
        QueryRequest? gotQueryRequest = null;
        fakeDb.QueryAsync(Arg.Any<QueryRequest>(), Arg.Any<CancellationToken>())
            .Returns(new QueryResponse
            {
                HttpStatusCode = HttpStatusCode.OK,
                ScannedCount = 1,
                Items = new List<Dictionary<string, AttributeValue>>
                {
                    new Dictionary<string, AttributeValue>
                    {
                        ["PK"] = new AttributeValue
                        {
                            S = "item-1"
                        },
                        ["SK"] = new AttributeValue
                        {
                            S = "item-details"
                        },
                        ["Name"] = new AttributeValue
                        {
                            S = "test"
                        }
                    }
                }
            }).AndDoes(x => gotQueryRequest = x.Arg<QueryRequest>());

        var sut = DynamoDbWrapperBuilder.Build(fakeDb);
        var rsp = await sut.Query().PartitionKey().Equals("item-1")
            .SortKey().Equals("item-details")
            .Fetch<TestRecordWithReservedWord>();

        gotQueryRequest.ShouldMatchApproved();
    }
}

public record TestRecordWithReservedWord
{
    public string PK { get; init; }
    public string SK { get; init; }
    public string Name { get; init; }
}

