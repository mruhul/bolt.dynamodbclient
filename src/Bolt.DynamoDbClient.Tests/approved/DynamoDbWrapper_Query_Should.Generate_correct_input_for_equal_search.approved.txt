{
  "gotRsp": {
    "items": [
      {
        "pk": "book-1",
        "sk": "details",
        "title": "book-title",
        "publishedAt": "2022-12-31T13:00:00Z",
        "price": 100.34,
        "identifier": "8359fc3f-b5ee-4622-9993-2b5cfae03a1e",
        "publishStatus": "Draft"
      }
    ],
    "count": 1,
    "lastEvaluatedKey": {},
    "scannedCount": 1
  },
  "gotRequest": {
    "attributesToGet": [],
    "consistentRead": false,
    "expressionAttributeNames": {
      "#SK": "SK",
      "#PK": "PK"
    },
    "expressionAttributeValues": {
      ":PK": {
        "bool": false,
        "isBOOLSet": false,
        "bs": [],
        "isBSSet": false,
        "l": [],
        "isLSet": false,
        "m": {},
        "isMSet": false,
        "ns": [],
        "isNSSet": false,
        "null": false,
        "s": "partition-key",
        "ss": [],
        "isSSSet": false
      },
      ":SK": {
        "bool": false,
        "isBOOLSet": false,
        "bs": [],
        "isBSSet": false,
        "l": [],
        "isLSet": false,
        "m": {},
        "isMSet": false,
        "ns": [],
        "isNSSet": false,
        "null": false,
        "s": "sort-key",
        "ss": [],
        "isSSSet": false
      }
    },
    "keyConditionExpression": "#PK = :PK and #SK = :SK",
    "keyConditions": {},
    "limit": 50,
    "isLimitSet": true,
    "projectionExpression": "PK, SK, Title, PublishedAt, Price, Identifier, PublishStatus",
    "queryFilter": {},
    "scanIndexForward": false,
    "tableName": "test-store"
  }
}